### DISCUSSION TITLE
Memory Efficient Clustering

### DISCUSSION THREAD

**Discussion by DanielRichardBiber** [rdkit/rdkit#5107] [Q&A]
Looking to do clustering for diversity purposes for >10M molecules but after a short foray into the available RDKit options, found that my distance matrix is just too big for memory in most machines.
What techniques are generally used for this sort of problem?

- Is clustering performed on a smaller subset (~50K) and then the remaining molecules are sorted in?
- Are there clustering algorithms that don't need random access to the entire matrix?
- Is sklearn used with bitwise fingerprints as there are options for memory efficient algorithms (mini-batch K-means)?

**Comment by greglandrum** [comment]
The RDKit has a couple of algorithms for doing this built in. 
This blog post describes one of them (sphere excusion clustering) and links to more information about the other (the MaxMin picker).
Both algorithms can operate directly on vectors of bit vectors and don't need to generate the distance matrix.
